ddl/index.go

```diff
@@ -966,112 +984,86 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
-
-	if err != nil {
-		// update the reorg handle that has been processed.
-		err1 := kv.RunInNewTxn(reorgInfo.d.store, true, func(txn kv.Transaction) error {
-			return errors.Trace(reorgInfo.UpdateReorgMeta(txn, nextHandle, reorgInfo.EndHandle, reorgInfo.PhysicalTableID))
-		})
-		metrics.BatchAddIdxHistogram.WithLabelValues(metrics.LblError).Observe(elapsedTime)
-		log.Warnf("[ddl-reorg] total added index for %d rows, this task [%d,%d) add index for %d failed %v, take time %v, update handle err %v",
-			*totalAddedCount, startHandle, nextHandle, taskAddedCount, err, elapsedTime, err1)
+		elapsedTime := time.Since(startTime).Seconds()
+		if task.result.err != nil {
+			cancel()
+			// update the reorg handle that has been processed.
+			err1 := kv.RunInNewTxn(reorgInfo.d.store, true, func(txn kv.Transaction) error {
+				return errors.Trace(reorgInfo.UpdateReorgMeta(txn, task.result.nextHandle, reorgInfo.EndHandle, reorgInfo.PhysicalTableID))
+			})
+			metrics.BatchAddIdxHistogram.WithLabelValues(metrics.LblError).Observe(elapsedTime)
+			log.Warnf("[ddl-reorg] total added index for %d rows, this task [%d,%d) add index for %d failed %v, take time %v, update handle err %v",
+				totalAddedCount, task.startHandle, task.result.nextHandle, task.result.addedCount, task.result.err, elapsedTime, err1)
+			return errors.Trace(task.result.err)
+		}
+		totalAddedCount += int64(task.result.addedCount)
+		w.reorgCtx.setNextHandle(task.result.nextHandle)
+		metrics.BatchAddIdxHistogram.WithLabelValues(metrics.LblOK).Observe(elapsedTime)
+		log.Infof("[ddl-reorg] total added index for %d rows, this task [%d,%d) added index for %d rows, take time %v",
+			totalAddedCount, task.startHandle, task.result.nextHandle, task.result.addedCount, elapsedTime)
+	}
+	// check err for close doingTaskCh first.
+	select {
+	case err := <-errCh:
 		return errors.Trace(err)
+	default:
 	}
-
-	// nextHandle will be updated periodically in runReorgJob, so no need to update it here.
-	w.reorgCtx.setNextHandle(nextHandle)
-	metrics.BatchAddIdxHistogram.WithLabelValues(metrics.LblOK).Observe(elapsedTime)
-	log.Infof("[ddl-reorg] total added index for %d rows, this task [%d,%d) added index for %d rows, take time %v",
-		*totalAddedCount, startHandle, nextHandle, taskAddedCount, elapsedTime)
 	return nil
 }
 
 // sendRangeTaskToWorkers sends tasks to workers, and returns remaining kvRanges that is not handled.
-func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker, reorgInfo *reorgInfo, totalAddedCount *int64, kvRanges []kv.KeyRange) ([]kv.KeyRange, error) {
-	batchTasks := make([]*reorgIndexTask, 0, len(workers))
-	physicalTableID := reorgInfo.PhysicalTableID
-
+func sendRangeTaskToWorkers(ctx context.Context, t table.Table, reorgInfo *reorgInfo, kvRanges []kv.KeyRange, availableWorkerCh chan *addIndexWorker, doingTaskCh chan *reorgIndexTask) error {
 	// Build reorg indices tasks.
 	for _, keyRange := range kvRanges {
+		select {
+		case <-ctx.Done():
```

Comment:
Seems like this context will never be canceled.

Reply:
Maybe cancel in `waitTaskFinish`.